### ISSUE TITLE
[BUG] BKCommonLib (and TrainCarts) doesnt load on Magma

### ISSUE THREAD

**Issue by vico93** [magmafoundation/Magma#202]
As i suspected, BKCommonLib (and every dependent plugins) doesnt load with Magma.

**To Reproduce**
1. Download latest Magma release (Magma-f05bb82-server);
2. Download latest [BKCommonLib](https://www.spigotmc.org/resources/bkcommonlib.39590/) (used `BKCommonLib-1.16.1-v1.jar` here) and [TrainCarts](https://www.spigotmc.org/resources/traincarts.39592/) (used here `TrainCarts-1.16.1-v1.jar` - although both says 1.16.1 in filename berkerkiller states both mods are backward-compatible with 1.12.2)
3. Load Magma
4. Get [this output](https://hastebin.com/fobarobilu.md) for debug.log

**Expected behavior**
BKCommonLib and TrainCarts should load sucessfully on Magma

**Info (please complete the following information):**
- Mod Name: [BKCommonLib](https://github.com/bergerhealer/BKCommonLib)
 - Mod Version: 1.16.1-v1 (latest on spigot page)
- Mod Name: [TrainCarts](https://github.com/bergerhealer/TrainCarts)
 - Mod Version: 1.16.1-v1 (latest on spigot page)
 - Magma Version: f05bb82
 - OS: Windows 10 1909 x64

**Additional context**
I already ran on [this similar problem back with Mohist](https://github.com/bergerhealer/BKCommonLib/issues/91), and at the time @bbayu123 said:

> I believe we can't do anything about it, adding compatibility for it would require a complete remap of the NMS calls, which comprise a large amount of the plugin. In layman's terms, the whole plugin will need to be rewritten.

Although later @bergerkiller himself said he would look at compatibility if discord members of his server voted in favour of it. Since i doubt this will happen i hope something could be done on Magma's end meanwhile.

**Comment by Hexeption** [comment]
This is being worked on by the plugin developer

**Comment by vico93** [comment]
But the refereed issue was marked as "wontfix", plus the wording of the first comment on there, so i thought no support for any kind of Forge/Spigot hybrid

**Comment by bbayu123** [comment]
It depends on whether we are able to receive the mapping used in your hybrid implementation. In the past, cauldron developers and all forks out there did not provide the mapping they used. Since BK and TrainCarts heavily depends on direct NMS calls, this proves to be a problem. And even if they did provide it, redirecting all existing calls would be a huge task and takes a lot of time.
@bergerkiller maybe you can provide more information and correct me if I'm wrong?

**Comment by vico93** [comment]
> It depends on whether we are able to receive the mapping used in your hybrid implementation.  In the past, cauldron developers and all forks out there did not provide the mapping they used.

@Hexeption  is the case of Magma too?

**Comment by Hexeption** [comment]
Our mapping are opensource they have access to them in the ’src/main/resources/mappings’

**Comment by bbayu123** [comment]
Yea, will need to check with the actual developer as to what's hindering the progress, and I might be wrong as to what the reason is.

**Comment by vico93** [comment]
> Yea, will need to check with the actual developer as to what's hindering the progress, and I might be wrong as to what the reason is.

@bbayu123 well @Hexeption is the actual developer of magma. He just replied my previous quote...

**Comment by bbayu123** [comment]
> > Yea, will need to check with the actual developer as to what's hindering the progress, and I might be wrong as to what the reason is.
> 
> @bbayu123 well @Hexeption is the actual developer of magma. He just replied my previous quote...
Nah i'm talking about TrainCarts

**Comment by vico93** [comment]
Ahh sorry for that

**Comment by bbayu123** [comment]
Ok just got word, he is currently writing an adapter logic to support magma and mohist, but once that's done, "_there's a lot of work remaining to support the subtle differences in method names_".
He also mentioned that he doesn't know "_what name differences are due to forge, and which are due to mohist/magma_"
So... there's that.

**Comment by vico93** [comment]
thanks for the info @bbayu123

**Comment by vico93** [comment]
Closing because @bergerkiller actually added support for hybrid servers on his side:

> So it's taken some months, but there is now rudimentary support for Mohist (and Magma/CatServer) in BKCommonLib. Feel free to check it out, and report back any issues you may find. There are some warnings at startup which can be ignored.
